### ISSUE TITLE
Implement Factory Design Pattern with Registry `get_object(string)`

### ISSUE THREAD

**Issue by mdsrqbl** [sign-language-translator/sign-language-translator#42]
#### Description

Let's improve the current factory-like functions in this library by implementing a factory design pattern with a registry. This enhancement will enable users to extend factory functions with their products. Below are some rough ideas and possible directions to achieve this goal:

#### Ideas

1. **Decorator for Registering Classes or Functions**:
   - Implement a decorator to register string short codes with classes for factory functions.
   - Example: See [`matplotlib.animation.MovieWriterRegistry`](https://github.com/matplotlib/matplotlib/blob/725ee995000985b2ee24d1b21cd777e0811272c8/lib/matplotlib/animation.py#L95), where [`@writers.register('pillow')`](https://github.com/matplotlib/matplotlib/blob/725ee995000985b2ee24d1b21cd777e0811272c8/lib/matplotlib/animation.py#L496) is added above a class.
   - Allow users to register their custom classes and functions.
   - Include an option in the decorator to take a list of misspellings or alternative names.
   - Detect and avoid name conflicts.

2. **String Names**:
 Currently All string literals are centralized in `slt.config.enums` and current objects return that value from a static method called `.name()` which the factory functions use.
   - Evaluate the use of frozen dataclasses instead of Enum classes if they offer better performance or clarity. (In that case, should we rename `enums.py` to `constants.py` or something else?)
   - How should we group all these literals for the ease of users. Should it be done by making a separate registry for each category of objects or left as it currently is (enum of Set[str] `slt.ModelCodeGroups.ALL_VIDEO_EMBEDDING_MODELS`)?

3. **Refactor Current Factory Functions**:
   - Identify and refactor existing factory functions to use the new registry-based design.
   - Current functions include:
     - `slt.models.get_model(name: str) -> torch.nn.Module`
     - `slt.languages.get_text_language(name: str) -> TextLanguage`
     - `slt.languages.get_sign_language(name: str) -> SignLanguage`
     - `slt.vision.get_sign_wrapper_class(name: str) -> Type[Sign]`
     - `slt.config.enums.normalize_short_code(name) -> str`

#### Task List

- [ ] Design and implement the decorator for registering classes & functions.
- [ ] Design an abstract factory and singleton registry/registries.
- [ ] Handle string normalization and removed models.
- [ ] Refactor existing factory functions to align with the new design.

#### Notes

- These are preliminary ideas meant to guide the implementation. The assignee is encouraged to explore and determine the best mechanism to achieve the outlined goals.
- The aim is to create a robust, maintainable, and user-friendly factory design pattern with a registry system that enables users to customize and integrate this library.
